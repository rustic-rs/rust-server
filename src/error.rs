//! Error types

use abscissa_core::error::{BoxError, Context};
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use std::{
    fmt::{self, Display},
    io,
    ops::Deref,
    result::Result,
};

pub type AppResult<T> = Result<T, Error>;
pub type ApiResult<T> = Result<T, ApiErrorKind>;

/// Kinds of errors
#[derive(Clone, Debug, Eq, thiserror::Error, PartialEq)]
pub enum ErrorKind {
    /// Error in configuration file
    #[error("config error")]
    Config,

    /// Input/output error
    #[error("I/O error")]
    Io,

    /// General storage error
    #[error("storage error")]
    GeneralStorageError,

    /// Missing user input
    #[error("missing user input")]
    MissingUserInput,
}

#[derive(Debug, thiserror::Error, displaydoc::Display)]
pub enum ApiErrorKind {
    /// Internal server error: `{0}`
    InternalError(String),
    /// Bad request: `{0}`
    BadRequest(String),
    /// Filename `{0}` not allowed
    FilenameNotAllowed(String),
    /// Path `{0}` is ambiguous with internal types and not allowed
    AmbiguousPath(String),
    /// Path `{0}` not allowed
    PathNotAllowed(String),
    /// Path `{0}` is not valid
    InvalidPath(String),
    /// Path `{0}` is not valid unicode
    NonUnicodePath(String),
    /// Creating directory failed: `{0}`
    CreatingDirectoryFailed(String),
    /// Not yet implemented
    NotImplemented,
    /// File not found: `{0}`
    FileNotFound(String),
    /// Fetting file metadata failed: `{0}`
    GettingFileMetadataFailed(String),
    /// Range not valid
    RangeNotValid,
    /// Seeking file failed
    SeekingFileFailed,
    /// Multipart range not implemented
    MultipartRangeNotImplemented,
    /// General range error
    GeneralRange,
    /// Conversion from length to u64 failed
    ConversionToU64Failed,
    /// Opening file failed: `{0}`
    OpeningFileFailed(String),
    /// Writing file failed: `{0}`
    WritingToFileFailed(String),
    /// Finalizing file failed: `{0}`
    FinalizingFileFailed(String),
    /// Getting file handle failed
    GettingFileHandleFailed,
    /// Removing file failed: `{0}`
    RemovingFileFailed(String),
    /// Reading from stream failed
    ReadingFromStreamFailed,
    /// Removing repository folder failed: `{0}`
    RemovingRepositoryFailed(String),
    /// Bad authentication header
    AuthenticationHeaderError,
    /// Failed to authenticate user: `{0}`
    UserAuthenticationError(String),
    /// General Storage error: `{0}`
    GeneralStorageError(String),
}

impl IntoResponse for ApiErrorKind {
    fn into_response(self) -> Response {
        let response = match self {
            ApiErrorKind::InternalError(err) => (
                StatusCode::INTERNAL_SERVER_ERROR,
                format!("Internal server error: {}", err),
            ),
            ApiErrorKind::BadRequest(err) => (
                StatusCode::BAD_REQUEST,
                format!("Internal server error: {}", err),
            ),
            ApiErrorKind::FilenameNotAllowed(filename) => (
                StatusCode::FORBIDDEN,
                format!("filename {filename} not allowed"),
            ),
            ApiErrorKind::AmbiguousPath(path) => (
                StatusCode::FORBIDDEN,
                format!("path {path} is ambiguous with internal types and not allowed"),
            ),
            ApiErrorKind::PathNotAllowed(path) => {
                (StatusCode::FORBIDDEN, format!("path {path} not allowed"))
            }
            ApiErrorKind::NonUnicodePath(path) => (
                StatusCode::BAD_REQUEST,
                format!("path {path} is not valid unicode"),
            ),
            ApiErrorKind::InvalidPath(path) => {
                (StatusCode::BAD_REQUEST, format!("path {path} is not valid"))
            }
            ApiErrorKind::CreatingDirectoryFailed(err) => (
                StatusCode::INTERNAL_SERVER_ERROR,
                format!("error creating dir: {:?}", err),
            ),
            ApiErrorKind::NotImplemented => (
                StatusCode::NOT_IMPLEMENTED,
                "not yet implemented".to_string(),
            ),
            ApiErrorKind::FileNotFound(path) => {
                (StatusCode::NOT_FOUND, format!("file not found: {path}"))
            }
            ApiErrorKind::GettingFileMetadataFailed(err) => (
                StatusCode::INTERNAL_SERVER_ERROR,
                format!("error getting file metadata: {err}"),
            ),
            ApiErrorKind::RangeNotValid => (StatusCode::BAD_REQUEST, "range not valid".to_string()),
            ApiErrorKind::SeekingFileFailed => (
                StatusCode::INTERNAL_SERVER_ERROR,
                "error seeking file".to_string(),
            ),
            ApiErrorKind::MultipartRangeNotImplemented => (
                StatusCode::NOT_IMPLEMENTED,
                "multipart range not implemented".to_string(),
            ),
            ApiErrorKind::ConversionToU64Failed => (
                StatusCode::INTERNAL_SERVER_ERROR,
                "error converting length to u64".to_string(),
            ),
            ApiErrorKind::OpeningFileFailed(err) => (
                StatusCode::INTERNAL_SERVER_ERROR,
                format!("error opening file: {err}"),
            ),
            ApiErrorKind::WritingToFileFailed(err) => (
                StatusCode::INTERNAL_SERVER_ERROR,
                format!("error writing file: {err}"),
            ),
            ApiErrorKind::FinalizingFileFailed(err) => (
                StatusCode::INTERNAL_SERVER_ERROR,
                format!("error finalizing file: {err}"),
            ),
            ApiErrorKind::GettingFileHandleFailed => (
                StatusCode::INTERNAL_SERVER_ERROR,
                "error getting file handle".to_string(),
            ),
            ApiErrorKind::RemovingFileFailed(err) => (
                StatusCode::INTERNAL_SERVER_ERROR,
                format!("error removing file: {err}"),
            ),
            ApiErrorKind::GeneralRange => {
                (StatusCode::INTERNAL_SERVER_ERROR, "range error".to_string())
            }
            ApiErrorKind::ReadingFromStreamFailed => (
                StatusCode::INTERNAL_SERVER_ERROR,
                "error reading from stream".to_string(),
            ),
            ApiErrorKind::RemovingRepositoryFailed(err) => (
                StatusCode::INTERNAL_SERVER_ERROR,
                format!("error removing repository folder: {:?}", err),
            ),
            ApiErrorKind::AuthenticationHeaderError => (
                StatusCode::FORBIDDEN,
                "Bad authentication header".to_string(),
            ),
            ApiErrorKind::UserAuthenticationError(err) => (
                StatusCode::FORBIDDEN,
                format!("Failed to authenticate user: {:?}", err),
            ),
            ApiErrorKind::GeneralStorageError(err) => (
                StatusCode::INTERNAL_SERVER_ERROR,
                format!("Storage error: {:?}", err),
            ),
        };

        response.into_response()
    }
}

impl ErrorKind {
    /// Create an error context from this error
    pub fn context(self, source: impl Into<BoxError>) -> Context<ErrorKind> {
        Context::new(self, Some(source.into()))
    }
}

/// Error type
#[derive(Debug)]
pub struct Error(Box<Context<ErrorKind>>);

impl Deref for Error {
    type Target = Context<ErrorKind>;

    fn deref(&self) -> &Context<ErrorKind> {
        &self.0
    }
}

impl Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0.fmt(f)
    }
}

impl std::error::Error for Error {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        self.0.source()
    }
}

impl From<ErrorKind> for Error {
    fn from(kind: ErrorKind) -> Self {
        Context::new(kind, None).into()
    }
}

impl From<Context<ErrorKind>> for Error {
    fn from(context: Context<ErrorKind>) -> Self {
        Error(Box::new(context))
    }
}

impl From<io::Error> for Error {
    fn from(err: io::Error) -> Self {
        ErrorKind::Io.context(err).into()
    }
}
