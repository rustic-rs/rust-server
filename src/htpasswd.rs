use std::{
    collections::{hash_map::Entry, HashMap},
    fmt::{Display, Formatter},
    fs::{self, read_to_string},
    io::Write,
    path::PathBuf,
};

use htpasswd_verify::md5::{format_hash, md5_apr1_encode};
use rand::{distributions::Alphanumeric, thread_rng, Rng};

use crate::error::{ApiErrorKind, ApiResult, AppResult, ErrorKind};
use abscissa_core::SecretString;

pub mod constants {
    pub(super) const SALT_LEN: usize = 8;
}

#[derive(Clone, Debug)]
pub struct Htpasswd {
    pub path: PathBuf,
    pub credentials: HashMap<String, Credential>,
}

impl Htpasswd {
    pub fn from_file(pth: &PathBuf) -> ApiResult<Htpasswd> {
        let mut c: HashMap<String, Credential> = HashMap::new();
        if pth.exists() {
            read_to_string(pth)
                .map_err(|err| {
                    ApiErrorKind::InternalError(format!(
                        "Could not read htpasswd file: {} at {:?}",
                        err, pth
                    ))
                })?
                .lines() // split the string into an iterator of string slices
                .map(String::from) // make each slice into a string
                .for_each(|line| match Credential::from_line(line) {
                    None => {}
                    Some(cred) => {
                        let _ = c.insert(cred.name.clone(), cred);
                    }
                })
        }
        Ok(Htpasswd {
            path: pth.clone(),
            credentials: c,
        })
    }

    pub fn users(&self) -> Vec<String> {
        self.credentials.keys().cloned().collect()
    }

    pub fn create(&mut self, name: &str, pass: &str) -> AppResult<()> {
        let cred = Credential::new(name, pass);

        let _ = self.insert(cred)?;

        Ok(())
    }

    pub fn read(&self, name: &str) -> Option<&Credential> {
        self.credentials.get(name)
    }

    /// Update can be used for both new, and existing credentials
    pub fn update(&mut self, name: &str, pass: &str) -> AppResult<()> {
        let cred = Credential::new(name, pass);

        let _ = self.insert(cred)?;

        Ok(())
    }

    /// Removes one credential by username
    pub fn delete(&mut self, name: &str) -> Option<Credential> {
        self.credentials.remove(name)
    }

    pub fn insert(&mut self, cred: Credential) -> AppResult<Credential> {
        let result = self.credentials.entry(cred.name.clone());

        match result {
            Entry::Occupied(mut entry) => Ok(entry.insert(cred)),
            Entry::Vacant(entry) => {
                return Err(ErrorKind::Io
                    .context(format!(
                        "Entry already exists, could not insert credential: `{}`. Please use update instead.",
                        entry.key()
                    ))
                    .into());
            }
        }
    }

    pub fn to_file(&self) -> ApiResult<()> {
        let mut file = fs::OpenOptions::new()
            .create(true)
            .truncate(false)
            .write(true)
            .open(&self.path)
            .map_err(|err| {
                ApiErrorKind::OpeningFileFailed(format!(
                    "Could not open htpasswd file: {} at {:?}",
                    err, self.path
                ))
            })?;

        for (_n, c) in self.credentials.iter() {
            let _e = file.write(c.to_line().as_bytes()).map_err(|err| {
                ApiErrorKind::WritingToFileFailed(format!(
                    "Could not write to htpasswd file: {} at {:?}",
                    err, self.path
                ))
            });
        }
        Ok(())
    }
}

#[derive(Clone, Debug)]
pub struct Credential {
    name: String,
    hash_val: Option<String>,
    pw: Option<SecretString>,
}

impl Credential {
    pub fn new(name: &str, pass: &str) -> Self {
        let salt: String = thread_rng()
            .sample_iter(&Alphanumeric)
            .take(constants::SALT_LEN)
            .map(char::from)
            .collect();
        let hash = md5_apr1_encode(pass, salt.as_str());
        let hash = format_hash(hash.as_str(), salt.as_str());

        Credential {
            name: name.into(),
            hash_val: Some(hash),
            pw: Some(pass.into()),
        }
    }

    /// Returns a credential struct from a htpasswd file line
    pub fn from_line(line: String) -> Option<Credential> {
        let spl: Vec<&str> = line.split(':').collect();
        if !spl.is_empty() {
            return Some(Credential {
                name: spl.first().unwrap().to_string(),
                hash_val: Some(spl.get(1).unwrap().to_string()),
                pw: None,
            });
        }
        None
    }

    pub fn to_line(&self) -> String {
        if self.hash_val.is_some() {
            format!(
                "{}:{}\n",
                self.name.as_str(),
                self.hash_val.as_ref().unwrap()
            )
        } else {
            "".into()
        }
    }
}

impl Display for Credential {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        writeln!(f, "Struct: Credential")?;
        writeln!(f, "\tUser: {}", self.name.as_str())?;
        writeln!(f, "\tHash: {}", self.hash_val.as_ref().unwrap())?;
        if self.pw.is_none() {
            writeln!(f, "\tPassword: None")?;
        } else {
            writeln!(f, "\tPassword: {:?}", &self.pw.as_ref().unwrap())?;
        }
        Ok(())
    }
}

#[cfg(test)]
mod test {
    use crate::auth::{Auth, AuthChecker};
    use crate::htpasswd::Htpasswd;
    use anyhow::Result;
    use insta::assert_debug_snapshot;
    
    use std::path::Path;

    #[test]
    fn test_htpasswd() -> Result<()> {
        let htpasswd_path = Path::new("tests/fixtures/test_data");
        let htpasswd_file = htpasswd_path.join(".htpasswd");

        let mut ht = Htpasswd::from_file(&htpasswd_file)?;

        assert_debug_snapshot!(ht);

        let _ = ht.update("Administrator", "stuff");
        let _ = ht.update("backup-user", "its_me");

        assert_debug_snapshot!(ht);

        let auth = Auth::from_file(false, &htpasswd_file)?;
        assert!(auth.verify("Administrator", "stuff"));
        assert!(auth.verify("backup-user", "its_me"));

        Ok(())
    }
}
