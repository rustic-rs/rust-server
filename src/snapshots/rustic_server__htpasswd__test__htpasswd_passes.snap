---
source: src/htpasswd.rs
expression: htpasswd
---
path = ''
[credentials.Administrator]
name = 'Administrator'
hash = '[hash]'

[credentials.backup-user]
name = 'backup-user'
hash = '[hash]'
